cirq-core/cirq/circuits/circuit.py

```diff
@@ -2841,8 +2832,44 @@ def _group_until_different(items: Iterable[_TIn], key: Callable[[_TIn], _TKey],
     return ((k, [val(i) for i in v]) for (k, v) in itertools.groupby(items, key))
 
 
+def _group_into_moment_compatible(inputs: Sequence[_MOMENT_OR_OP]) -> Iterator[List[_MOMENT_OR_OP]]:
+    """Groups sequential ops into those that can coexist in a single moment.
+
+    This function will go through the input sequence in order, emitting lists of sequential
+    operations that can go into a single moment. It does not try to rearrange the elements or try
+    to move them to open slots in earlier moments; it simply processes them in order and outputs
+    them. i.e. the output, if flattened, will equal the input.
+
+    Actual Moments in the input will always be emitted by themselves as a single-element list.
+
+    Examples:
+        [X(a), X(b), X(a)] -> [[X(a), X(b)], [X(a)]]
+        [X(a), X(a), X(b)] -> [[X(a)], [X(a), X(b)]]
+        [X(a), Moment(X(b)), X(c)] -> [[X(a)], [Moment(X(b))], [X(c)]]"""
+    i = 0
+    batch: List[_MOMENT_OR_OP] = []
+    while i < len(inputs):
+        if isinstance(inputs[i], Moment):
+            yield [inputs[i]]
+            i += 1
+            continue
+        batch_qubits: Set['cirq.Qid'] = set()
+        while i < len(inputs):
+            mop = inputs[i]
+            qs = mop.qubits
+            if isinstance(mop, Moment) or not batch_qubits.isdisjoint(qs):
+                yield batch
+                batch = []
+                break
+            batch.append(mop)
+            batch_qubits.update(qs)
+            i += 1
+    if batch:
+        yield batch
```

Comment:
Nit - consider this variation to avoid a nested loop.

```suggestion
    """
    batch: List[_MOMENT_OR_OP] = []
    batch_qubits: Set['cirq.Qid'] = set()
    for mop in inputs:
        is_moment = isinstance(mop, cirq.Moment)
        if (is_moment and batch) or not batch_qubits.isdisjoint(mop.qubits):
            yield batch
            batch = []
            batch_qubits.clear()
        if is_moment:
            yield [mop]
            continue
        batch.append(mop)
        batch_qubits.update(mop.qubits)
    if batch:
        yield batch
```

Reply:
Nice, I like this better too.